tls/s2n_server_hello.c

```diff
@@ -76,14 +87,31 @@ int s2n_server_hello_recv(struct s2n_connection *conn)
         if (s2n_is_hello_retry_req(conn)) {
             GUARD(s2n_server_hello_retry_recv(conn));
         }
+
         /* Check echoed session ID matches */
         S2N_ERROR_IF(session_id_len != conn->session_id_len || memcmp(session_id, conn->session_id, session_id_len), S2N_ERR_BAD_MESSAGE);
         conn->actual_protocol_version = conn->server_protocol_version;
         GUARD(s2n_set_cipher_as_client(conn, cipher_suite_wire));
     } else {
         uint8_t actual_protocol_version;
-
         conn->server_protocol_version = (uint8_t)(protocol_version[0] * 10) + protocol_version[1];
+
+        /* Detect downgrade attacks according to RFC 8446 section 4.1.3 */
```

Comment:
Probably worth putting this in its own function to keep recv clean(er)

Reply:
Good call, refactored in next push.